---
title: "CellPhy - SC-Caller conversion"
author: 'Alexey Kovlov, [João M Alves*](mailto:jmfernandesalves@gmail.com), Alexandros Stamatakis & David Posada'
date: "July 2020"
output:
  pdf_document:
    number_sections: yes
  html_document:
    highlight: textmate
    keep_md: yes
    number_sections: yes
    theme: cosmo
    toc: yes
---

<br/><br/>  

# **Introductory Note**
<p style='text-align: justify;'>In the following supporting document, we will show how to easily transform a VCF file generated by the [**_SC-Caller_**](https://github.com/biosinodx/SCcaller/) (a single-cell specific variant caller) into a [**CellPhy-GL**](https://github.com/amkozlov/raxml-ng/tree/cellphy) readable VCF. Before we start, users should make sure that both _BCFtools_ and _Tabix_ from [**Samtools**](http://www.htslib.org/) are installed and available.</p>    

***
# **VCF file conversion**
<p style='text-align: justify;'>SC-Caller VCF files can be converted by running the _sc-caller-convert.sh_ script as follows:</p>
\scriptsize
```{r, engine = 'bash', eval = FALSE}

$ ./sc-caller-convert.sh 
Usage: ./sc-caller-convert.sh inputVCF SamplePrefix
Created by: Alexey Kovlov, Joao M Alves, Alexandros Stamatakis & David Posada - June 2020

$ ./sc-caller-convert.sh CRC_Cell-A.22.sccaller.vcf Cell-A
```
\normalsize

<p style='text-align: justify;'>If everything went as expected, you should have generated a new VCF (**Cell-A.PL-fixed.vcf**) that is now ready for downstream analysis. Below we provide a detailed explanation of which changes are being made to the original VCF file.</p>  

## **Step-by-step explanation**
<p style='text-align: justify;'>While many callers currently being used with single-cell genomic data generate VCF files with a standard "PL" field (e.g., [GATK-HaplotypeCaller](https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller), [Monovar](https://github.com/KChen-lab/MonoVar), [Prosolo](https://github.com/ProSolo/prosolo)), it is important to highlight that the "PL" definition may differ from its standard meaning in different tools. Indeed, [SC-Caller](https://github.com/biosinodx/SCcaller) for instance uses the "PL" field to store not only the likelihood of heterozygous and alternative homozygous genotypes, but also the likelihood of sequencing noise and amplification artifacts.    </p>

<p style='text-align: justify;'>Below we show an example of a VCF generated using SC-Caller (version 2.0.0).  </p>

\scriptsize
```{r, engine = 'bash', eval = FALSE}
$ grep -v "#" CRC_Cell-A.22.sccaller.vcf | head -n 5
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	CELL001
22	16149851	.	G	A	13	.	NS=1	GT:SO:AD:BI:GQ:PL	0/0:NA:0,6:0.6:13:91,54,13,0
22	16190307	.	A	-1C	4	.	NS=1	GT:SO:AD:BI:GQ:PL	0/0:NA:0,1:0.6:4:15,9,2,0
22	16193737	.	A	C,G	150	multiple-genotype	NS=1	GT:SO:AD:BI:GQ:PL	0/0:True:317,1:0.993:150:91,227,64,11570
22	16195864	.	G	A	150	.	NS=1	GT:SO:AD:BI:GQ:PL	0/0:True:136,1:0.6:150:39,93,310,3479
```
\normalsize

<p style='text-align: justify;'>Looking at this output, you should notice that:

* SC-Caller renames the single-cell to "CELL001";
* The VCF contains calls other than single-nucleotide variants (i.e., SNVs);
* The "PL" field of **bi-allelic** sites is composed of 4 entries, as opposed to the 3 values usually observed in VCF files that were generated following the [_format specifications_*](https://samtools.github.io/hts-specs/VCFv4.3.pdf);
* The "PL" field will always contain 4 entries, regardless of the amount of alternative variants detected.

As a consequence, we will need to rename the sample to its proper ID and trim the VCF to exclude all indels and non-biallelic positions (as we won't be able to get the likelihood scores for all possible genotypes). This can be easily done using BCFtools.  </p>

\scriptsize

```{r, engine = 'bash', eval = FALSE}
# Rename sample in VCF
$ cat temp_rename
CELL001 Cell-A
$ bcftools reheader -s temp_rename CRC_Cell-A.22.sccaller.vcf -o temp.renamed.vcf

# Remove indels and non-biallelic sites
$ bcftools view --types snps -f "." temp.renamed.vcf -o temp.snvs.vcf
```

\normalsize

<p style='text-align: justify;'>Once this is done, we should end up with a trimmed VCF that solely contains bi-allelic sites:</p>

\scriptsize

```{r, engine = 'bash', eval = FALSE}
$ grep -v "##" temp.snvs.vcf | head -n 5
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	Cell-A
22	16149851	.	G	A	13	.	NS=1	GT:SO:AD:BI:GQ:PL	0/0:NA:0,6:0.6:13:91,54,13,0
22	16195864	.	G	A	150	.	NS=1	GT:SO:AD:BI:GQ:PL	0/0:True:136,1:0.6:150:39,93,310,3479
22	16195889	.	T	C	150	.	NS=1	GT:SO:AD:BI:GQ:PL	0/0:True:136,1:0.6:150:34,89,306,4496
22	16195900	.	A	G	150	.	NS=1	GT:SO:AD:BI:GQ:PL	0/0:True:132,1:0.6:150:34,87,298,4273
```

\normalsize

<p style='text-align: justify;'>Afterwards, we need to transform the "PL" field at each site into the standard "PL" format. Following [**SC-caller authors’ suggestions**](https://github.com/biosinodx/SCcaller/issues/10), we take the highest likelihood score of the first two values (i.e., sequencing noise, amplification artifact) as the phred-scaled genotype likelihood of the reference homozygous (0/0) genotype, and the remaining values as the likelihood for heterozygous (0/1) and alternative homozygous (1/1) genotypes, respectively. </p>

\textcolor{red}{\textbf{Cautionary remark on SC-Caller developers suggestion:}}  
\fbox{%
\begin{minipage}{6.5in}
It is perhaps important to mention that the SC-Caller authors suggest to take "the bigger number as the PL combined" as the 0/0 genotype likelihood. We interpreted this suggestion as to take the one with the "the highest likelihood" (which in this case would be the smallest integer value).
\end{minipage}}  

<br>  

<p style='text-align: justify;'>Our custom script will then essentially rename the previous "PL" field as "FPL" and add a standard "PL" field to our VCF to make it CellPhy readable:</p>
\scriptsize
```{r, engine = 'bash', eval = FALSE}
$ grep -v "##" temp.PL-fixed.vcf | head
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	Cell-A
22	16149851	.	G	A	13	.	NS=1	GT:SO:AD:BI:GQ:FPL:PL	0/0:NA:0,6:0.6:13:91,54,13,0:54,13,0
22	16195864	.	G	A	150	.	NS=1	GT:SO:AD:BI:GQ:FPL:PL	0/0:True:136,1:0.6:150:39,93,310,3479:39,310,3479
22	16195889	.	T	C	150	.	NS=1	GT:SO:AD:BI:GQ:FPL:PL	0/0:True:136,1:0.6:150:34,89,306,4496:34,306,4496
22	16195900	.	A	G	150	.	NS=1	GT:SO:AD:BI:GQ:FPL:PL	0/0:True:132,1:0.6:150:34,87,298,4273:34,298,4273
```
\normalsize

*** 

# **Merging VCF files for downstream analyses**

<p style='text-align: justify;'>Once all single-cell VCF files are converted, we can easily merge them into a multi-sample VCF using BCFtools.</p>

\scriptsize

```{r, engine = 'bash', eval = FALSE}
$ for i in *PL-fixed.vcf
  do
  bgzip $i
  tabix -p vcf ${i}.gz
  done
$ ls *.vcf.gz > listMERGE
$ bcftools merge -l listMERGE -O vcf -o CRC-allCells.vcf
```
\normalsize

<p style='text-align: justify;'>The VCF is now ready for CellPhy (or for any additional filtering steps).</p>
